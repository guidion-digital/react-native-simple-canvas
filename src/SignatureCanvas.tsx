import React, { forwardRef, RefObject, useCallback, useImperativeHandle, useMemo, useRef, useState, MutableRefObject } from 'react';
import { PanResponder, StyleSheet, View, ViewStyle } from 'react-native';
import { Path, Svg } from 'react-native-svg';

import { Point } from './interfaces/Point';
import { spline } from './helpers/signature';

interface SignatureCanvasProps {
  onDragEvent?: () => void;
  onSignatureChange?: (isEmpty: boolean) => void;
  strokeColor?: string;
  strokeWidth?: number;
  backgroundColor?: string;
  style?: ViewStyle;
  minPoints?: number;
}

export interface SignatureCanvasRef {
  resetImage: () => void;
  getSVG: () => RefObject<Svg>;
  isEmpty: () => boolean;
  getPoints: () => Point[];
  setPoints: (points: Point[]) => void;
}

export const clearCanvas = (ref: MutableRefObject<SignatureCanvasRef | null>) => {
  ref.current?.resetImage();
}; 

export const SignatureCanvas = forwardRef<SignatureCanvasRef, SignatureCanvasProps>(
  ({ 
    onDragEvent,
    onSignatureChange,
    strokeColor = 'black',
    strokeWidth = 3,
    backgroundColor = 'transparent',
    style,
    minPoints = 2
  }, ref) => {
    const [paths, setPaths] = useState<string[]>([]);
    const pointsRef = useRef<Point[]>([]);
    const svgRef = useRef(null);
    const isDrawing = useRef(false);

    const canvasRef = useRef<View>(null);
    const canvasOffset = useRef({ x: 0, y: 0 });

    const addPoint = useCallback((point: Point) => {
      pointsRef.current.push(point);
      return true;
    }, []);

    const resetImage = useCallback(() => {
      setPaths([]);
      pointsRef.current = [];
      onSignatureChange?.(true);
    }, [onSignatureChange]);

    const isEmpty = useCallback(() => {
      return paths.length === 0;
    }, [paths]);

    const getPoints = useCallback(() => {
      return [...pointsRef.current];
    }, []);

    const setPoints = useCallback((newPoints: Point[]) => {
      if (newPoints.length < minPoints) return;
      
      pointsRef.current = newPoints;
      const path = spline(newPoints, 1, false);
      setPaths([path]);
      onSignatureChange?.(false);
    }, [minPoints, onSignatureChange]);

    const onLayout = useCallback(() => {
      canvasRef.current?.measureInWindow((x, y) => {
        canvasOffset.current = { x, y };
      });
    }, []);

    const panResponder = useMemo(
      () =>
        PanResponder.create({
          onStartShouldSetPanResponder: () => true,
          onMoveShouldSetPanResponder: () => true,
          onPanResponderGrant: (event) => {
            const { pageX, pageY } = event.nativeEvent;
            const point = {
              x: pageX - canvasOffset.current.x,
              y: pageY - canvasOffset.current.y
            };
            isDrawing.current = true;
            pointsRef.current = [point];
            setPaths(prevPaths => [...prevPaths, `M ${point.x} ${point.y}`]);
            onDragEvent?.();
          },
          onPanResponderMove: (event) => {
            if (!isDrawing.current) return;
            const { pageX, pageY } = event.nativeEvent;
            const point = {
              x: pageX - canvasOffset.current.x,
              y: pageY - canvasOffset.current.y
            };

            if (addPoint(point)) {
              if (pointsRef.current.length >= minPoints) {
                const path = spline(pointsRef.current, 1, false);
                setPaths(prevPaths => {
                  const newPaths = [...prevPaths];
                  newPaths[newPaths.length - 1] = path;
                  return newPaths;
                });
                onSignatureChange?.(false);
              }
            }
          },
          onPanResponderRelease: () => {
            isDrawing.current = false;
          },
        }),
      [onDragEvent, addPoint, minPoints, onSignatureChange]
    );

    useImperativeHandle(
      ref,
      () => ({
        getSVG: () => svgRef,
        resetImage,
        isEmpty,
        getPoints,
        setPoints
      }),
      [resetImage, isEmpty, getPoints, setPoints]
    );

    const pathElements = useMemo(
      () =>
        paths.map((path, index) => (
          <Path
            key={index}
            d={path}
            stroke={strokeColor}
            strokeWidth={strokeWidth}
            fill="none"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        )),
      [paths, strokeColor, strokeWidth]
    );

    return (
      <View style={[styles.container, style]}>
        <View
          ref={canvasRef}
          onLayout={onLayout}
          style={[styles.signatureBox, { backgroundColor }]}
          {...panResponder.panHandlers}
        >
          <Svg ref={svgRef} height="100%" width="100%">
            {pathElements}
          </Svg>
        </View>
      </View>
    );
  }
);

SignatureCanvas.displayName = 'SignatureCanvas';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  signatureBox: {
    flex: 1,
    borderWidth: 1,
    borderColor: '#ccc',
  },
});
