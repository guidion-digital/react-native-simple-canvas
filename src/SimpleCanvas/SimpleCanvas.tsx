import React, {
  forwardRef,
  MutableRefObject,
  RefObject,
  useCallback,
  useImperativeHandle,
  useMemo,
  useRef,
  useState
} from 'react';
import { PanResponder, StyleSheet, View, ViewStyle } from 'react-native';
import { Path, Svg } from 'react-native-svg';

import { Point } from '../interfaces/Point';
import { spline } from '../helpers/signature';

interface SimpleCanvasProps {
  onDragEvent?: () => void;
  onCanvasChange?: (isEmpty: boolean) => void;
  strokeColor?: string;
  strokeWidth?: number;
  backgroundColor?: string;
  style?: ViewStyle;
  minPoints?: number;
}

export interface SimpleCanvasRef {
  resetImage: () => void;
  getSVG: () => RefObject<Svg>;
  isEmpty: () => boolean;
  getPoints: () => Point[];
  setPoints: (points: Point[]) => void;
}

export const clearCanvas = (ref: MutableRefObject<SimpleCanvasRef | null>) => {
  ref.current?.resetImage();
};

export const SimpleCanvas = forwardRef<SimpleCanvasRef, SimpleCanvasProps>(
  ({
    onDragEvent,
    onCanvasChange,
    strokeColor = 'black',
    strokeWidth = 3,
    backgroundColor = 'transparent',
    style,
    minPoints = 2
  }, ref) => {
    const [paths, setPaths] = useState<string[]>([]);
    const pointsRef = useRef<Point[]>([]);
    const svgRef = useRef(null);
    const isDrawing = useRef(false);

    const canvasRef = useRef<View>(null);
    const canvasOffset = useRef({ x: 0, y: 0 });

    const addPoint = useCallback((point: Point) => {
      pointsRef.current.push(point);
      return true;
    }, []);

    const resetImage = useCallback(() => {
      setPaths([]);
      pointsRef.current = [];
      onCanvasChange?.(true);
    }, [onCanvasChange]);

    const isEmpty = useCallback(() => {
      return paths.length === 0;
    }, [paths]);

    const getPoints = useCallback(() => {
      return [...pointsRef.current];
    }, []);

    const setPoints = useCallback((newPoints: Point[]) => {
      if (newPoints.length < minPoints) return;

      pointsRef.current = newPoints;
      const path = spline(newPoints, 1, false);
      setPaths([path]);
      onCanvasChange?.(false);
    }, [minPoints, onCanvasChange]);

    const onLayout = useCallback(() => {
      canvasRef.current?.measureInWindow((x, y) => {
        canvasOffset.current = { x, y };
      });
    }, []);

    const panResponder = useMemo(
      () =>
        PanResponder.create({
          onStartShouldSetPanResponder: () => true,
          onMoveShouldSetPanResponder: () => true,
          onPanResponderGrant: (event) => {
            const { pageX, pageY } = event.nativeEvent;
            const point = {
              x: pageX - canvasOffset.current.x,
              y: pageY - canvasOffset.current.y
            };
            isDrawing.current = true;
            pointsRef.current = [point];
            setPaths(prevPaths => [...prevPaths, `M ${point.x} ${point.y}`]);
            onDragEvent?.();
          },
          onPanResponderMove: (event) => {
            if (!isDrawing.current) return;
            const { pageX, pageY } = event.nativeEvent;
            const point = {
              x: pageX - canvasOffset.current.x,
              y: pageY - canvasOffset.current.y
            };

            if (addPoint(point)) {
              if (pointsRef.current.length >= minPoints) {
                const path = spline(pointsRef.current, 1, false);
                setPaths(prevPaths => {
                  const newPaths = [...prevPaths];
                  newPaths[newPaths.length - 1] = path;
                  return newPaths;
                });
                onCanvasChange?.(false);
              }
            }
          },
          onPanResponderRelease: () => {
            isDrawing.current = false;
          },
        }),
      [onDragEvent, addPoint, minPoints, onCanvasChange]
    );

    useImperativeHandle(
      ref,
      () => ({
        getSVG: () => svgRef as unknown as RefObject<Svg>,
        resetImage,
        isEmpty,
        getPoints,
        setPoints
      }),
      [resetImage, isEmpty, getPoints, setPoints]
    );

    const pathElements = useMemo(
      () =>
        paths.map((path, index) => (
          <Path
            key={index}
            d={path}
            stroke={strokeColor}
            strokeWidth={strokeWidth}
            fill="none"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        )),
      [paths, strokeColor, strokeWidth]
    );

    return (
      <View style={[styles.container, style]}>
        <View
          ref={canvasRef}
          onLayout={onLayout}
          style={[styles.signatureBox, { backgroundColor }]}
          {...panResponder.panHandlers}
        >
          <Svg ref={svgRef} height="100%" width="100%">
            {pathElements}
          </Svg>
        </View>
      </View>
    );
  }
);

SimpleCanvas.displayName = 'SimpleCanvas';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  signatureBox: {
    flex: 1,
    borderWidth: 1,
    borderColor: '#ccc',
  },
});
