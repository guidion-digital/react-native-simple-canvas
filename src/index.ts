import { clearCanvas, SimpleCanvas, SimpleCanvasRef } from './SimpleCanvas/SimpleCanvas';

export { clearCanvas, SimpleCanvas, SimpleCanvasRef };
