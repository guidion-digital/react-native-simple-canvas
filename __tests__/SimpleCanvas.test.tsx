import React, { createRef } from 'react';
import { SimpleCanvas, SimpleCanvasRef } from '../src/SimpleCanvas';
import renderer, { act, create } from 'react-test-renderer';
import { View } from 'react-native';

const mockTouchEvent = {
  nativeEvent: {
    pageX: 192,
    pageY: 124,
    touches: [{ pageX: 192, pageY: 124 }]
  },
  touchHistory: {
    indexOfSingleActiveTouch: 1,
    mostRecentTimeStamp: 411733262.676947,
    numberActiveTouches: 1,
    touchBank: [
      undefined,
      {
        currentPageX: 192,
        currentPageY: 124,
        currentTimeStamp: 411733262.676947,
        previousPageX: 192,
        previousPageY: 324,
        previousTimeStamp: 411733234.333552,
        startPageX: 202,
        startPageY: 647.6666564941406,
        startTimeStamp: 411731079.78750896,
        touchActive: true,
      },
    ],
  },
};

const mockMoveEvent = {
  ...mockTouchEvent,
  nativeEvent: {
    ...mockTouchEvent.nativeEvent,
    pageX: 200,
    pageY: 200,
  },
};

const mockOutOfContainerMoveEvent = {
  ...mockTouchEvent,
  nativeEvent: {
    ...mockTouchEvent.nativeEvent,
    pageX: 333,
    pageY: 333,
  },
};

describe('SimpleCanvas', () => {
  it('should render', () => {
    const tree = renderer.create(
      <SimpleCanvas onDragEvent={() => {}} onCanvasChange={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls onDragEvent when starting to draw', () => {
    const onDragEvent = jest.fn();
    const component = create(<SimpleCanvas onDragEvent={onDragEvent} />);
    const views = component.root.findAllByType(View);
    const panHandler = views[1];

    act(() => {
      panHandler.props.onLayout({ nativeEvent: { layout: { height: 300, width: 300 } } });
      panHandler.props.onStartShouldSetResponder();
      panHandler.props.onResponderGrant(mockTouchEvent);
    });

    act(() => panHandler.props.onResponderMove(mockMoveEvent));

    act(() => panHandler.props.onResponderRelease({}));

    expect(onDragEvent).toHaveBeenCalled();
  });

  it('stops drawing when out of container', () => {
    const onDragEvent = jest.fn();
    const component = create(<SimpleCanvas onDragEvent={onDragEvent} />);
    const views = component.root.findAllByType(View);
    const panHandler = views[1];

    act(() => {
      panHandler.props.onLayout({ nativeEvent: { layout: { height: 300, width: 300 } } });
      panHandler.props.onStartShouldSetResponder();
      panHandler.props.onResponderGrant(mockTouchEvent);
    });

    act(() => panHandler.props.onResponderMove(mockOutOfContainerMoveEvent));

    expect(onDragEvent).toHaveBeenCalled();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('resets the signature', () => {
    const ref = createRef<SimpleCanvasRef>();
    const component = create(<SimpleCanvas ref={ref} />);

    act(() => {
      const instance = component.root;
      const views = instance.findAllByType(View);
      views[1].props.onResponderGrant(mockTouchEvent);
    });

    expect(ref.current).not.toBeNull();

    act(() => {
      ref.current?.resetImage();
    });

    const instance = component.root;
    const paths = instance.findAllByProps({ stroke: 'black' });
    expect(paths.length).toBe(0);
  });
});
