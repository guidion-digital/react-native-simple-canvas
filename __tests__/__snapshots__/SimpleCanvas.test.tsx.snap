// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleCanvas should render 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    onLayout={[Function]}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      [
        {
          "borderColor": "#ccc",
          "borderWidth": 1,
          "flex": 1,
        },
        {
          "backgroundColor": "transparent",
        },
      ]
    }
  >
    <RNSVGSvgView
      bbHeight="100%"
      bbWidth="100%"
      focusable={false}
      height="100%"
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": "100%",
            "width": "100%",
          },
        ]
      }
      width="100%"
    >
      <RNSVGGroup
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
      />
    </RNSVGSvgView>
  </View>
</View>
`;

exports[`SimpleCanvas stops drawing when out of container 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    onLayout={[Function]}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      [
        {
          "borderColor": "#ccc",
          "borderWidth": 1,
          "flex": 1,
        },
        {
          "backgroundColor": "transparent",
        },
      ]
    }
  >
    <RNSVGSvgView
      bbHeight="100%"
      bbWidth="100%"
      focusable={false}
      height="100%"
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": "100%",
            "width": "100%",
          },
        ]
      }
      width="100%"
    >
      <RNSVGGroup
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGPath
          d="M 192 124 C 215.5 158.83333333333334 309.5 298.1666666666667 333 333"
          fill={null}
          propList={
            [
              "fill",
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth={3}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
